Generic mapper: add 'lazy' map mode to match room names only, ignoring exits

wiploo opened the pull request:
Easy room detection for games that don't allow a clear mode of detect closed/opened/door.

<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions
Now there is a way to allow mapper to better create room even if door are not correctly configured. Sometimes can't detect room with same name, but you can disable it if your mud allow to detect room.

In future PR i think that we can add a method to automatically detect and set exits with door, but for now keep it simple

wiploo committed:
Enable lazy_room_detect

Easy room detection for games that don't allow a clear mode of detect closed/opened/door.

wiploo committed:
Fixed missing function

Copy/paste error

vadi2 commented on src/mudlet-lua/lua/generic-mapper/generic_mapper.xml:
```diff
@@ -1729,6 +1736,18 @@ local function check_room(roomID, name, exits)
+local function check_room_only_name(roomID, name)
+   -- check to see if room name and exits match expecations
+    if not roomID then
+        error("Check Room Error: No ID",2)
```
```suggestion
        error("check_room_only_name: no roomID given",2)
```

vadi2 reviewed (commented):
Semantically I think what you're adding is a new mapping mode:

![image](https://user-images.githubusercontent.com/110988/103218250-d36c4e80-491a-11eb-9c8b-495380d1661e.png)

How about the new `lazy` one?

wiploo committed:
Create a mode_move to better handle mapper movement

wiploo commented:
Hi, we have tryed to do better and create a "mode_move" parameters to handle how generic_mapper interacts with already mapped rooms. generci_mapper "mode" is about how create new rooms.

wiploo commented:
PS: I haven't changed some file in 3rd-party tools, but github told so... I don't know how to revert to right versione because I can't find the right version. is it the development branch?

vadi2 commented:
Yeah it's in the development branch. If you want I can help reset it for you:

![image](https://user-images.githubusercontent.com/110988/103751719-07293180-5009-11eb-93ad-6127295c62d4.png)

vadi2 commented:
> Now there is a way to allow mapper to better create room even if door are not correctly configured.

Aren't we talking about a mapping mode though here?

wiploo commented:
> Yeah it's in the development branch. If you want I can help reset it for you:
> 
> ![image](https://user-images.githubusercontent.com/110988/103751719-07293180-5009-11eb-93ad-6127295c62d4.png)

Yes, please.

wiploo commented:
> > Now there is a way to allow mapper to better create room even if door are not correctly configured.
> 
> Aren't we talking about a mapping mode though here?

No, the mapping mode is about creating a room if not already present in mudlet. move mode is about detecting a room while moving. This "move mode" can work with normal/simple/complex map mode and improve theme.

vadi2 committed:
Restore edbee

vadi2 committed:
Update to latest dbsqd

vadi2 commented:
I've fixed the submodules for you.

>map move &lt;normal, easy, or door&gt;&lt;/link&gt; - Sets the movement mode, which
            defines how mapper interact with rooms already mapped. Can detect and create doors automatically.        

Sorry but you're confusing me. The move mode is also changing the map, ie mapping?

wiploo commented:
> I've fixed the submodules for you.
> 
> > map move <normal, easy, or door></link> - Sets the movement mode, which
> > defines how mapper interact with rooms already mapped. Can detect and create doors automatically.
> 
> Sorry but you're confusing me. The move mode is also changing the map, ie mapping?

Move mode only changes how a room is detected when you came back to a room already mapped. If a room is not detected mapper use normale/simple/complex mode to create the room, otherwise the "move mode" is used to create/ignore the room exits of the destination room.

So you can have:
- mode simple and move easy
- mode normal and move door
- mode complex and move door (useless probably)
- etc.

wiploo commented:
> I've fixed the submodules for you.

Many thanks!

wiploo commented:
> I've fixed the submodules for you.
> 
> > map move <normal, easy, or door></link> - Sets the movement mode, which
> > defines how mapper interact with rooms already mapped. Can detect and create doors automatically.
> 
> Sorry but you're confusing me. The move mode is also changing the map, ie mapping?

I have tryed, all this options are viable (plus the normal)

- simple e easy
- simple e door
- normal e easy
- normale e door
- complex e easy
- complex e door

wiploo commented:
@Edru2 @Eraene @Kebap @keneanung @SlySven (and any other mudlet developers) can someone help understand the better way to implement this PR?

vadi2 commented:
https://github.com/Mudlet/Mudlet/pull/4530#issuecomment-756030036 yeah that is a problem. It's hard to claim it is user-friendly and simple to use when there's 8 different permutation of settings you need to get right before the mapper will work OK.

That's why I am still asking if move move changes the map or not, to see if we can make this something that is acceptable for players. I know for your players you can just tell "turn on this and that option" and it solves the problem, but it also needs to work games that don't have Mudlet-specific instructions.

> Move mode only changes how a room is detected when you came back to a room already mapped. If a room is not detected mapper use normale/simple/complex mode to create the room, otherwise the "move mode" is used to create/ignore the room exits of the destination room.

Hmm this starts saying a move mode does not change the map and then it says move mode creates exits, so it does change the map? Can we fold the option into existing move modes somehow or make it auto-detect when it needs to use this option?

The player experience is important here, thanks for your patience :smiley:

wiploo commented:
Ok, i can change the "move_mode" to a "mode", but some question:
- how can i handle the autocreation of door? somone may not like it, so we need additional config?
- I'd like to have the opinion of some map users to understand if the really like mode_move or prefear a simple, but less flexible "mode"

Thanks for the reply

vadi2 commented:
I don't understand the part about doors - it's not well explained here or in the PR - but isn't the move mode you're proposing already covered by the earlier added `custom_name_search`?

It won't auto-create doors and etc, but then again the move mode has nothing to do with map mode, so further thinking would needed there

wiploo commented:
I'll try to explain:
- custom_name_search changes how the room name is detected in room description. Sometimes the default method is too complex or inaccurate, so a custom function works better
- this PR is about two new functionallity:
  a) if the "easy" move_mode is active, when you enter a room, generic_mapper check if a room is already mapped only searching by room name. Without this PR, generic_mapper search a room using room name and visibile exits so sometimes it fails when there area door not detected the first time the room is mapped (for example because MUD don't show them as doors)
  b) if the "door" move_mode si active, the room search is performed like in "easy" move_mode, but if a room already mapped is detected, eventually new exits or old exits on that room compared to the saved one, are automatically marked as door

I hope that description is more understandable

Eraene commented:
I still don't understand what problem this PR is trying to solve. If you are talking about hidden doors, that is solved by setting exits as doors in the map UI. What is this solving that setting exits as doors in the map UI does not?

wiploo commented:
I have two answers for your objections:
- with the current generic_mapper code, if I enter a room that is already mapped but has different exits not marked as door, a new room is created. In our game is not always clear that an exits is a door so mistakes are common during mapping and users have to delete and re-arrange rooms and exits many times
- like you told I can manually set every exits as a door, but maybe, I can configure generic_mapper to do that for me or help me in this operation. With the move_mode "easy" I can configure generic_mapper to match room only by their name so if there is a not marked exit with a door, no wrong room is created

Summary:
- I'd like to have an option to detect room already mapped only by their room name (not checking exits)
- I'd like to have an option to auto-mark exits as door when I came back to a room already mapped if new/missing rooms are discovered

I hope this can help understand why I make this PR.

vadi2 commented:
Sorry for the delay in reviewing this.

>    I'd like to have an option to detect room already mapped only by their room name (not checking exits)

The `custom_name_search` can do that already for you, right?

>    I'd like to have an option to auto-mark exits as door when I came back to a room already mapped if new/missing rooms are discovered

The mapper raises an `onNewRoom` event. You could hook into that to auto-mark all exits as doors. Not keen on having an option that does this because I really don't see anyone else wanting every single exit to be shown as a door on the map.

Would that work for you?

wiploo commented:
> Sorry for the delay in reviewing this.

Nope, we all have personal live, or we should :-D

> 
> > I'd like to have an option to detect room already mapped only by their room name (not checking exits)
> 
> The `custom_name_search` can do that already for you, right?

Unfortunatelly no, `custom_name_search` extract the name of the room from mudlet output in a custom way. With all mapping mode and even with normal `name_search, the room name is used WITH room exits to detect if the new room is already in database or is a new one. I need to match room ONLY using room name, ignoring exits

> > I'd like to have an option to auto-mark exits as door when I came back to a room already mapped if new/missing rooms are discovered
> 
> The mapper raises an `onNewRoom` event. You could hook into that to auto-mark all exits as doors. Not keen on having an option that does this because I really don't see anyone else wanting every single exit to be shown as a door on the map.
> 
> Would that work for you?

Unfortunatelly no, because our MUD don't show if an exit is a door in the room description. I need to compare new detected exits and the old ones, to find door. The `onNewRoom` event is not applicabile because I don't have the old data, maybe a `onOldRoom` event is what I need, but I think that isn't the right solution. Door detection code can be shared among every mud because they all have exits.

vadi2 commented:
> the room name is used WITH room exits to detect if the new room is already in database or is a new one. I need to match room ONLY using room name, ignoring exits

OK, I see what you mean. It does sound like you need a new `map mode` option here.

> Door detection code can be shared among every mud because they all have exits.

Other games list their doors though, they don't turn into dissapearing exits. As far as I know this is unique, maybe @Eraene you've ran into this before?

> The onNewRoom event is not applicabile because I don't have the old data

You can use the `onNewRoom` to keep track of what room you are in and that way you'll know what the old room was, and that way you can compare both rooms. Right?

wiploo commented:
Ok, we work on a new "mode" and we use the event "onMove" to handle auto door creation.

We come back soon with an update PR

vadi2 commented:
Thanks and sorry it took a month to sort out, hopefully next ones will be as quick as they were before!

wiploo committed:
create a new map mode

reverted all move_mode code, and created a new map mode to handle check_room only by name

wiploo commented:
Hey, let's try this one! :-D

wiploo committed:
versioning

vadi2 committed:
Merge remote-tracking branch 'origin/development' into lazy-exit-detection

vadi2 committed:
Minor wording fixes

vadi2 reviewed (commented):
Functionality looks good to me :) but we're thinking that 'basic' and 'simple' are gonna get confusing, thinking on a better name for you

wiploo commented:
Mhhh, it's not easy choose a good name compare to "simple", "normal" and "complex". I think that "basic" can be understand as a simpler way than "simple" mode. Maybe "name" or "onlyName" can be be more exaustive?

Unfortunatelly, I don't know english enought to choose a proper wording.

vadi2 commented:
I was thinking onlyname too but @Kebap wasn't in for it. Thinking more...

wiploo commented:
I think that "basic" can be the best choice. I hope that anyone can understand that "basic" is simplier then "simple", "simple" is simplier thant "normal" and "complex".  I am quite sure that matching a room with less information that the room name is impossible.

vadi2 committed:
Rename to map mode basic to lazy

vadi2 commented:
"basic" and "simple" in English are super similar though, to be honest after a while it'll get confusing. Decided to go with your original suggestion of "lazy" :+1:

vadi2 commented:
Didn't fix up commit in this one, habit
